L1Trigger/TrackFindingTMTT/src/L1KalmanComb.cc

```diff
@@ -0,0 +1,1829 @@
+    for (int i = 0; i < K.GetNrows(); i++) {
+      tmp(i, i) = 1;
+    }
+    for (int i = 0; i < K.GetNrows(); i++) {
+      for (int j = 0; j < K.GetNcols(); j++) {
+        for (int k = 0; k < h.GetNcols(); k++) {
+          tmp(i, k) += -1 * K(i, j) * h(j, k);
+        }
+      }
+    }
+    new_xcov.Clear();
+    new_xcov.ResizeTo(pxcov.GetNrows(), pxcov.GetNcols());
+    for (int i = 0; i < tmp.GetNrows(); i++) {
+      for (int j = 0; j < tmp.GetNcols(); j++) {
+        for (int k = 0; k < pxcov.GetNcols(); k++) {
+          new_xcov(i, k) += tmp(i, j) * pxcov(j, k);
+        }
+      }
+    }
+  }
+
+  void L1KalmanComb::resetStates() {
+    for (unsigned int i = 0; i < state_list_.size(); i++) {
+      delete state_list_.at(i);
+    }
+    state_list_.clear();
+  }
+
+  const KalmanState *L1KalmanComb::mkState(const L1track3D &candidate,
+                                           unsigned skipped,
+                                           unsigned layer,
+                                           unsigned layerId,
+                                           const KalmanState *last_state,
+                                           const std::vector<double> &x,
+                                           const TMatrixD &pxx,
+                                           const TMatrixD &K,
+                                           const TMatrixD &dcov,
+                                           const StubCluster *stubCluster,
+                                           double chi2rphi,
+                                           double chi2rz) {
+    KalmanState *new_state = new KalmanState(candidate,
+                                             skipped,
+                                             layer,
+                                             layerId,
+                                             last_state,
+                                             x,
+                                             pxx,
+                                             K,
+                                             dcov,
+                                             stubCluster,
+                                             chi2rphi,
+                                             chi2rz,
+                                             this,
+                                             &getTrackParams);
+
+    if (chi2rphi + chi2rz == 0) {
+      double new_state_chi2rphi = 0., new_state_chi2rz = 0.;
+      this->calcChi2(*new_state, new_state_chi2rphi, new_state_chi2rz);
+      new_state->setChi2(new_state_chi2rphi, new_state_chi2rz);
+    }
+
+    state_list_.push_back(new_state);
+    return new_state;
+  }
+
+  std::vector<double> L1KalmanComb::residual(const StubCluster *stubCluster,
+                                             const std::vector<double> &x,
+                                             double candQoverPt) const {
+    std::vector<double> vd = d(stubCluster);  // Get (phi relative to sector, z) of hit.
```

Comment:
To what are you referring? Functions residuals() and d() return temporary variables, so they can't be returned by ref. (They could be returned as a function argument to achieve this, but it was felt this would make the code harder to understand, as function arguments contain both inputs & outputs).

Reply:
I assume @kpedro88 meant something like
```suggestion
    std::vector<double> const& vd = d(stubCluster);  // Get (phi relative to sector, z) of hit.
```
which avoid the copy if the function returns a reference.

On the other hand, if the function already returns a temporary by value, it shouldn't make any difference.

Reply:
left unchanged